File: src/app/pages/tasks_page.py
# 版权所有 © Silent X Craft Launcher Dev 开发团队
#
# Silent X Craft Launcher (SXCL) 是一款由 Silent X Craft Launcher Dev 团队开发，
# 隶属于 SilentCodeTeams 旗下，并由 SilentStudio 管理的 Minecraft 第三方启动器。
#
# Copyright © Silent X Craft Launcher Development Team
#
# Silent X Craft Launcher (SXCL) is a third-party Minecraft launcher developed
# by the Silent X Craft Launcher Dev team, operating under the management of
# SilentCodeTeams, and overseen by SilentStudio.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version, WITH the Additional Terms described
# in the LICENSE file accompanying this program.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""Tasks page — session-aware task monitor with status icons."""

from __future__ import annotations

from PySide6.QtCore import Qt, Signal, QTimer
from PySide6.QtWidgets import (
    QHBoxLayout, QVBoxLayout, QWidget, QPushButton,
)
from qfluentwidgets import (
    BodyLabel, ProgressBar, StrongBodyLabel,
)

from src.app.common.base_page import BasePage


class _StatusIcon(QWidget):
    """任务状态图标：进行中 = 旋转，完成 = ✓，失败 = ✕，悬停显示 × 删除."""

    STATE_RUNNING = 0
    STATE_DONE = 1
    STATE_FAILED = 2

    def __init__(self, parent=None):
        super().__init__(parent)
        self._state = self.STATE_RUNNING
        self._spin = 0
        self._timer = QTimer(self)
        self._timer.timeout.connect(self._tick)

        self.setFixedSize(24, 24)
        self._spin_chars = ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"]

        self._label = BodyLabel("⠋", self)
        self._label.setAlignment(Qt.AlignCenter)
        self._label.setStyleSheet("font-size: 16px;")
        layout = QHBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.addWidget(self._label)

    def set_state(self, state: int):
        self._state = state
        if state == self.STATE_RUNNING:
            self._timer.start(120)
            self._label.setStyleSheet("color: #1890ff; font-size: 16px;")
            self.setCursor(Qt.ArrowCursor)
        elif state == self.STATE_DONE:
            self._timer.stop()
            self._label.setText("✓")
            self._label.setStyleSheet("color: #52c41a; font-size: 16px; font-weight: bold;")
            self.setCursor(Qt.ArrowCursor)
        elif state == self.STATE_FAILED:
            self._timer.stop()
            self._label.setText("✕")
            self._label.setStyleSheet("color: #ff4d4f; font-size: 16px; font-weight: bold;")
            self.setCursor(Qt.PointingHandCursor)

    def _tick(self):
        self._spin = (self._spin + 1) % len(self._spin_chars)
        self._label.setText(self._spin_chars[self._spin])


class _DeleteBtn(QPushButton):
    """鼠标悬停时出现的红色 × 删除按钮."""

    def __init__(self, parent=None):
        super().__init__("×", parent)
        self.setFixedSize(20, 20)
        self.setVisible(False)
        self.setStyleSheet("""
            QPushButton { border: none; color: #ff4d4f; font-size: 18px; font-weight: bold;
                          background: transparent; }
            QPushButton:hover { color: #ff0000; }
        """)


class TaskCard(QWidget):
    """A single task item showing download/launch progress."""

    clicked = Signal(object)
    delete_requested = Signal(str)

    def __init__(self, task_id: str, title: str, parent=None):
        super().__init__(parent)
        self.task_id = task_id

        self.setFixedHeight(80)
        self.setAttribute(Qt.WA_StyledBackground, True)
        self.setCursor(Qt.PointingHandCursor)
        self.setStyleSheet("""
            TaskCard {
                background: rgba(128, 128, 128, 0.06);
                border-radius: 8px;
                margin: 2px 0;
            }
            TaskCard:hover {
                background: rgba(128, 128, 128, 0.12);
            }
        """)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(16, 12, 16, 12)

        # Row 1: icon + title + status + delete
        row1 = QHBoxLayout()
        self.icon = _StatusIcon(self)
        row1.addWidget(self.icon)
        row1.addSpacing(8)
        self.title_label = StrongBodyLabel(title, self)
        row1.addWidget(self.title_label)
        row1.addStretch()
        self.status_label = BodyLabel("准备中", self)
        self.status_label.setStyleSheet("color: #888;")
        row1.addWidget(self.status_label)
        row1.addSpacing(4)
        self.delete_btn = _DeleteBtn(self)
        self.delete_btn.clicked.connect(lambda: self.delete_requested.emit(self.task_id))
        row1.addWidget(self.delete_btn)
        layout.addLayout(row1)

        # Row 2: progress bar
        self.progress_bar = ProgressBar(self)
        self.progress_bar.setFixedHeight(4)
        self.progress_bar.setRange(0, 100)
        self.progress_bar.setValue(0)
        layout.addWidget(self.progress_bar)

        # Row 3: detail text
        self.detail_label = BodyLabel("", self)
        self.detail_label.setStyleSheet("color: #999; font-size: 12px;")
        layout.addWidget(self.detail_label)

        # Hover → show delete button (only for done/failed)
        self._hover_shows_delete = False

    def update_progress(self, value: int, status: str, detail: str = ""):
        self.progress_bar.setValue(value)
        self.status_label.setText(status)
        if detail:
            self.detail_label.setText(detail)

    def set_running(self):
        self.icon.set_state(_StatusIcon.STATE_RUNNING)
        self._hover_shows_delete = False
        self.delete_btn.setVisible(False)

    def set_done(self):
        self.icon.set_state(_StatusIcon.STATE_DONE)
        self._hover_shows_delete = False
        self.progress_bar.setValue(100)
        self.delete_btn.setVisible(False)

    def set_failed(self):
        self.icon.set_state(_StatusIcon.STATE_FAILED)
        self._hover_shows_delete = True

    def mousePressEvent(self, event):
        self.clicked.emit(self)

    def enterEvent(self, event):
        if self._hover_shows_delete:
            self.delete_btn.setVisible(True)
        super().enterEvent(event)

    def leaveEvent(self, event):
        self.delete_btn.setVisible(False)
        super().leaveEvent(event)


class TasksPage(BasePage):
    """任务页面 — 显示下载、启动等正在进行的任务."""

    def __init__(self, parent=None):
        super().__init__(title="任务", subtitle="查看和管理正在进行的操作", parent=parent)
        self._tasks: dict[str, TaskCard] = {}
        self._build_content()

    def _build_content(self):
        self._empty_label = BodyLabel("暂无进行中的任务", self.view)
        self._empty_label.setAlignment(Qt.AlignCenter)
        self._empty_label.setStyleSheet("color: #888; font-size: 14px; margin: 60px 0;")
        self.add_content(self._empty_label)

        self._task_container = QWidget(self.view)
        self._task_layout = QVBoxLayout(self._task_container)
        self._task_layout.setContentsMargins(0, 0, 0, 0)
        self._task_layout.setAlignment(Qt.AlignTop)
        self._task_container.setVisible(False)
        self.add_content(self._task_container)

        self.add_stretch()

    # ── Public API ────────────────────────────────────────────

    def add_or_update_task(
        self, task_id: str, title: str,
        progress: int = 0, status: str = "准备中", detail: str = "",
    ) -> TaskCard:
        if task_id in self._tasks:
            card = self._tasks[task_id]
            card.update_progress(progress, status, detail)
            return card

        card = TaskCard(task_id, title, self._task_container)
        card.update_progress(progress, status, detail)
        card.clicked.connect(lambda c: self._on_task_clicked(c))
        card.delete_requested.connect(self.remove_task)
        self._tasks[task_id] = card
        self._task_layout.addWidget(card)
        self._refresh_visibility()
        return card

    def remove_task(self, task_id: str):
        card = self._tasks.pop(task_id, None)
        if card:
            self._task_layout.removeWidget(card)
            card.deleteLater()
            self._refresh_visibility()

    def update_task(
        self, task_id: str, progress: int = 0,
        status: str = "", detail: str = "",
    ):
        card = self._tasks.get(task_id)
        if card:
            card.update_progress(progress, status, detail)

    def set_task_running(self, task_id: str):
        card = self._tasks.get(task_id)
        if card:
            card.set_running()

    def set_task_done(self, task_id: str):
        card = self._tasks.get(task_id)
        if card:
            card.set_done()
        # 完成后 3 秒自动移除
        QTimer.singleShot(3000, lambda: self.remove_task(task_id))

    def set_task_failed(self, task_id: str):
        card = self._tasks.get(task_id)
        if card:
            card.set_failed()

    def _refresh_visibility(self):
        has = len(self._tasks) > 0
        self._empty_label.setVisible(not has)
        self._task_container.setVisible(has)

    def _on_task_clicked(self, card: TaskCard):
        mw = self.window()
        if hasattr(mw, 'navigate_to_task'):
            mw.navigate_to_task(card.task_id)

File: src/ui/pages/tasks_page.cpp
/*
 * (C) Silent X Craft Launcher
 * Copyright by SilentStudio.
 * All rights reserved.
 */

#include "page_factory.h"

#include <QEnterEvent>
#include <QHBoxLayout>
#include <QMouseEvent>
#include <QPainter>
#include <QPainterPath>
#include <QPen>
#include <QPushButton>
#include <QTimer>
#include <QVBoxLayout>

#if defined(_MSC_VER)
#pragma warning(push, 0) // libqf 的头在 /W4 下不是零警告,整体静音(见 libqf.h 的说明)
#endif
#include "fluent/fluent_controls.h" // PushButton / InfoBar
#include "fluent/fluent_labels.h"   // BodyLabel / StrongBodyLabel / SubtitleLabel / TitleLabel
#include "fluent/fluent_progress.h" // ProgressBar
#include "fluent/fluent_scroll.h"   // ScrollArea
#if defined(_MSC_VER)
#pragma warning(pop)
#endif

#include "fluent_theme.h"

namespace sxcl::ui {

namespace {

// 令牌小工具
QString tokenCss(const QString &name) { return FluentTheme::instance().tokenText(name); }

// QSS 用的令牌字符串。实测(6.11):Qt 的 QSS 解析器**吃不下带空格的 rgba**
// ("rgba(255, 255, 255, 0.05)" 会被整条声明丢掉 → 卡片底色不画),而令牌字符串
// (theme.py) 里就是带空格的 —— 所以进 QSS 前先把 rgba(...) 里的空格去掉,
// 颜色值不变,渲染结果与 Python 的设计意图(半透明白)一致。
QString tokenQss(const QString &name) {
    QString s = tokenCss(name);
    if (s.startsWith(QLatin1String("rgba"), Qt::CaseInsensitive)) {
        const int l = s.indexOf(QLatin1Char('('));
        const int r = s.lastIndexOf(QLatin1Char(')'));
        if (l > 0 && r > l)
            s = s.left(l + 1) + s.mid(l + 1, r - l - 1).remove(QLatin1Char(' ')) + s.mid(r);
    }
    return s;
}

} // namespace

// ──────────────────────────────────────────────────────────────── _StatusIcon

// tasks_page.py:40-96 —— 任务状态图标:进行中 = 转圈,完成 = 对勾,失败 = 叉。
// **三个状态全部自绘**(用户 2026-09-22 与 09-26 两次点名:「emoji 去掉,改成 svg」)——
// 以前这里用 "⠋⠙⠹…" 盲文转轮和 "✓/✕" 两个字符:它们落到哪个字体由系统决定,
// 大小 / 颜色 / 基线都不受控(在别的机器上就是 emoji 的观感)。颜色仍取主题令牌。
class TaskStatusIcon : public QWidget {
    Q_OBJECT
public:
    enum State { Running = 0, Done = 1, Failed = 2 };

    explicit TaskStatusIcon(QWidget *parent = nullptr) : QWidget(parent) {
        setFixedSize(24, 24);
        m_timer = new QTimer(this);
        m_timer->setInterval(120);
        connect(m_timer, &QTimer::timeout, this, &TaskStatusIcon::tick);
        connect(&FluentTheme::instance(), &FluentTheme::changed, this, [this] { update(); });
        m_timer->start();
    }

    void setState(int state) {
        m_state = state;
        if (state == Running) {
            m_timer->start();
            setCursor(Qt::ArrowCursor);
        } else {
            m_timer->stop();
            setCursor(state == Failed ? Qt::PointingHandCursor : Qt::ArrowCursor);
        }
        update(); // 状态变了立刻重画,不等下一帧
    }

protected:
    void paintEvent(QPaintEvent *) override {
        QPainter painter(this);
        painter.setRenderHint(QPainter::Antialiasing);
        const QRectF box = QRectF(rect()).adjusted(3.5, 3.5, -3.5, -3.5);
        if (m_state == Running) { // 转圈:每 tick 转 30°,画 3/4 圈
            QPen pen(FluentTheme::instance().accent(), 2.0);
            pen.setCapStyle(Qt::RoundCap);
            painter.setPen(pen);
            painter.setBrush(Qt::NoBrush);
            painter.drawArc(box, m_spin * 30 * 16, 270 * 16);
            return;
        }
        const ThemeTokens &tokens = FluentTheme::instance().tokens();
        QPen pen(m_state == Failed ? tokens.danger : tokens.success, 2.0);
        pen.setCapStyle(Qt::RoundCap);
        pen.setJoinStyle(Qt::RoundJoin);
        painter.setPen(pen);
        painter.setBrush(Qt::NoBrush);
        QPainterPath path; // 完成 = 对勾;失败 = 叉
        if (m_state == Failed) {
            path.moveTo(box.left(), box.top());
            path.lineTo(box.right(), box.bottom());
            path.moveTo(box.right(), box.top());
            path.lineTo(box.left(), box.bottom());
        } else {
            path.moveTo(box.left(), box.center().y() + 1.0);
            path.lineTo(box.center().x() - 1.0, box.bottom());
            path.lineTo(box.right(), box.top());
        }
        painter.drawPath(path);
    }

private:
    void tick() { // tasks_page.py:94-96(帧推进;这里推的是自绘弧的角度)
        m_spin = (m_spin + 1) % 12;
        update();
    }

    QTimer *m_timer = nullptr;
    int m_state = Running;
    int m_spin = 0;
};

// ──────────────────────────────────────────────────────────────── _DeleteBtn

// tasks_page.py:99-115 —— 鼠标悬停时出现的红色叉号删除按钮
class TaskDeleteButton : public QPushButton {
    Q_OBJECT
public:
    explicit TaskDeleteButton(QWidget *parent = nullptr)
        : QPushButton(QStringLiteral("×"), parent) {
        setFixedSize(20, 20);
        setVisible(false);
        connect(&FluentTheme::instance(), &FluentTheme::changed, this,
                [this] { refreshTheme(); });
        refreshTheme();
    }

private:
    void refreshTheme() {
        const QString danger = tokenCss(QStringLiteral("danger"));
        setStyleSheet(QStringLiteral(
                          "QPushButton { border: none; color: %1; font-size: 18px;"
                          " font-weight: bold; background: transparent; }\n"
                          "QPushButton:hover { color: %1; }")
                          .arg(danger));
    }
};

// ──────────────────────────────────────────────────────────────── TaskCard

// tasks_page.py:118-207 —— 一张任务卡(下载/启动进度)
class TaskCard : public QWidget {
    Q_OBJECT
public:
    TaskCard(const QString &taskId, const QString &title, QWidget *parent = nullptr)
        : QWidget(parent), m_taskId(taskId) {
        setFixedHeight(80);
        setAttribute(Qt::WA_StyledBackground, true);
        setCursor(Qt::PointingHandCursor);
        // Python 用类名选择器 TaskCard { … };Qt 的 QSS 对**命名空间里的类**要求写成
        // sxcl--ui--TaskCard(冒号换成 --),既啰嗦又容易踩坑 —— 这里改成等价的
        // objectName 选择器 #taskCard,渲染结果与 Python 一致。
        setObjectName(QStringLiteral("taskCard"));

        auto *layout = new QVBoxLayout(this);
        layout->setContentsMargins(16, 12, 16, 12);
        layout->setSpacing(6); // Python 用样式默认值(实测 6)

        // Row 1: 图标 + 标题 + 状态 + 删除
        auto *row1 = new QHBoxLayout();
        row1->setSpacing(6);
        m_icon = new TaskStatusIcon(this);
        row1->addWidget(m_icon);
        row1->addSpacing(8);
        m_title = new StrongBodyLabel(title, this);
        row1->addWidget(m_title);
        row1->addStretch();
        m_status = new BodyLabel(QStringLiteral("准备中"), this);
        m_status->setStyleSheet(QStringLiteral("color: #888;"));
        row1->addWidget(m_status);
        row1->addSpacing(4);
        m_delete = new TaskDeleteButton(this);
        connect(m_delete, &QPushButton::clicked, this,
                [this] { emit deleteRequested(m_taskId); });
        row1->addWidget(m_delete);
        layout->addLayout(row1);

        // Row 2: 进度条
        m_progress = new ProgressBar(this);
        m_progress->setFixedHeight(4);
        m_progress->setRange(0, 100);
        m_progress->setValue(0);
        layout->addWidget(m_progress);

        // Row 3: 明细
        m_detail = new BodyLabel(QString(), this);
        m_detail->setStyleSheet(QStringLiteral("color: #999; font-size: 12px;"));
        layout->addWidget(m_detail);

        connect(&FluentTheme::instance(), &FluentTheme::changed, this,
                [this] { refreshTheme(); });
        refreshTheme();
    }

    QString taskId() const { return m_taskId; }

    void updateProgress(int value, const QString &status, const QString &detail = QString()) {
        m_progress->setValue(value);
        m_status->setText(status);
        if (!detail.isEmpty())
            m_detail->setText(detail);
    }

    void setRunning() { // tasks_page.py:182-185
        m_icon->setState(TaskStatusIcon::Running);
        m_hoverShowsDelete = false;
        m_delete->setVisible(false);
    }
    void setDone() { // tasks_page.py:187-191
        m_icon->setState(TaskStatusIcon::Done);
        m_hoverShowsDelete = false;
        m_progress->setValue(100);
        m_delete->setVisible(false);
    }
    void setFailed() { // tasks_page.py:193-195
        m_icon->setState(TaskStatusIcon::Failed);
        m_hoverShowsDelete = true;
    }

signals:
    void clicked(TaskCard *card);
    void deleteRequested(const QString &taskId);

protected:
    void mousePressEvent(QMouseEvent *event) override { // tasks_page.py:197-198
        emit clicked(this);
        QWidget::mousePressEvent(event);
    }
    void enterEvent(QEnterEvent *event) override { // tasks_page.py:200-203
        if (m_hoverShowsDelete)
            m_delete->setVisible(true);
        QWidget::enterEvent(event);
    }
    void leaveEvent(QEvent *event) override { // tasks_page.py:205-207
        m_delete->setVisible(false);
        QWidget::leaveEvent(event);
    }

private:
    void refreshTheme() { // tasks_page.py:169-174
        setStyleSheet(QStringLiteral(
                          "#taskCard { background: %1; border-radius: 8px;"
                          " margin: 2px 0; }\n"
                          "#taskCard:hover { background: %2; }")
                          .arg(tokenQss(QStringLiteral("hoverBg")),
                               tokenQss(QStringLiteral("hoverBgStrong"))));
    }

    QString m_taskId;
    TaskStatusIcon *m_icon = nullptr;
    StrongBodyLabel *m_title = nullptr;
    BodyLabel *m_status = nullptr;
    TaskDeleteButton *m_delete = nullptr;
    ProgressBar *m_progress = nullptr;
    BodyLabel *m_detail = nullptr;
    bool m_hoverShowsDelete = false;
};

// ──────────────────────────────────────────────────────────────── TasksPage

// tasks_page.py:210-293
class TasksPage : public ScrollArea {
    Q_OBJECT
public:
    explicit TasksPage(QWidget *parent = nullptr) : ScrollArea(parent) {
        setObjectName(QStringLiteral("TasksPage"));
        setWidgetResizable(true);
        setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);

        // 页面底色钉令牌 bg(#202020):抓参考图时 Python 也是这么钉的(grab_reference_ui.py:141
        //   page.setStyleSheet("QWidget { background: %s }" % token("bg")))——参考图的内容区
        // 就是 #202020,不是内容栈那层半透明白。栈自己的 rgba(255,255,255,0.0314) 只该在
        // 窗口左上圆角那半像素露出来(见 main_window.cpp 的圆角取证),页面不钉底色整片会变 #272727。
        setStyleSheet(QStringLiteral("QScrollArea { background: %1; }")
                          .arg(FluentTheme::instance().tokens().bg.name()));

        // BasePage 骨架(base_page.py:46-60)
        m_view = new QWidget(this);
        m_view->setStyleSheet(QStringLiteral("background: transparent;"));
        setWidget(m_view);
        m_box = new QVBoxLayout(m_view);
        m_box->setContentsMargins(28, 24, 28, 24);
        m_box->setSpacing(16);
        m_box->setAlignment(Qt::AlignTop);
        m_title = new TitleLabel(QStringLiteral("任务"), m_view);
        m_subtitle = new SubtitleLabel(QStringLiteral("查看和管理正在进行的操作"), m_view);
        m_subtitle->setTextColor(QColor(QStringLiteral("#606060")),
                                 QColor(QStringLiteral("#AAAAAA")));
        m_box->addWidget(m_title);
        m_box->addWidget(m_subtitle);

        buildContent();
    }
    // ---- 对外 API(tasks_page.py:235-283;主窗口登记任务时按 id 驱动)----

    Q_INVOKABLE TaskCard *addOrUpdateTask(const QString &taskId, const QString &title,
                                          int progress = 0,
                                          const QString &status = QStringLiteral("准备中"),
                                          const QString &detail = QString()) {
        if (TaskCard *existing = m_tasks.value(taskId, nullptr)) {
            existing->updateProgress(progress, status, detail);
            return existing;
        }
        auto *card = new TaskCard(taskId, title, m_container);
        card->updateProgress(progress, status, detail);
        connect(card, &TaskCard::clicked, this, [this](TaskCard *c) { onTaskClicked(c); });
        connect(card, &TaskCard::deleteRequested, this, [this](const QString &id) {
            removeTask(id);
        });
        m_tasks.insert(taskId, card);
        m_layout->addWidget(card);
        refreshVisibility();
        return card;
    }

    Q_INVOKABLE void removeTask(const QString &taskId) {
        TaskCard *card = m_tasks.value(taskId, nullptr);
        if (!card)
            return;
        m_tasks.remove(taskId);
        m_layout->removeWidget(card);
        card->deleteLater();
        refreshVisibility();
    }

    Q_INVOKABLE void updateTask(const QString &taskId, int progress = 0,
                                const QString &status = QString(),
                                const QString &detail = QString()) {
        if (TaskCard *card = m_tasks.value(taskId, nullptr))
            card->updateProgress(progress, status, detail);
    }

    Q_INVOKABLE void setTaskRunning(const QString &taskId) {
        if (TaskCard *card = m_tasks.value(taskId, nullptr))
            card->setRunning();
    }

    Q_INVOKABLE void setTaskDone(const QString &taskId) {
        if (TaskCard *card = m_tasks.value(taskId, nullptr))
            card->setDone();
        // 完成后 3 秒自动移除(tasks_page.py:277-278)
        QTimer::singleShot(3000, this, [this, taskId] { removeTask(taskId); });
    }

    Q_INVOKABLE void setTaskFailed(const QString &taskId) {
        if (TaskCard *card = m_tasks.value(taskId, nullptr))
            card->setFailed();
    }

    // 自检用(报告/测试):当前任务卡数量
    Q_INVOKABLE int taskCount() const { return int(m_tasks.size()); }

private:
    void buildContent() { // tasks_page.py:218-231
        m_emptyLabel = new BodyLabel(QStringLiteral("暂无进行中的任务"), m_view);
        m_emptyLabel->setAlignment(Qt::AlignCenter);
        m_emptyLabel->setStyleSheet(
            QStringLiteral("color: #888; font-size: 14px; margin: 60px 0;"));
        m_box->addWidget(m_emptyLabel);

        m_container = new QWidget(m_view);
        m_layout = new QVBoxLayout(m_container);
        m_layout->setContentsMargins(0, 0, 0, 0);
        m_layout->setSpacing(6); // Python 用样式默认值(实测 6)
        m_layout->setAlignment(Qt::AlignTop);
        m_container->setVisible(false);
        m_box->addWidget(m_container);

        m_box->addStretch(1);
    }

    void refreshVisibility() { // tasks_page.py:285-288
        const bool has = !m_tasks.isEmpty();
        m_emptyLabel->setVisible(!has);
        m_container->setVisible(has);
    }

    void onTaskClicked(TaskCard *card) { // tasks_page.py:290-293
        if (!card)
            return;
        // Python: mw.navigate_to_task(card.task_id) —— 主窗口没这个方法就什么也不做
        QMetaObject::invokeMethod(window(), "navigateToTask", Qt::DirectConnection,
                                  Q_ARG(QString, card->taskId()));
    }

    QWidget *m_view = nullptr;
    QVBoxLayout *m_box = nullptr;
    TitleLabel *m_title = nullptr;
    SubtitleLabel *m_subtitle = nullptr;
    BodyLabel *m_emptyLabel = nullptr;
    QWidget *m_container = nullptr;
    QVBoxLayout *m_layout = nullptr;
    QMap<QString, TaskCard *> m_tasks;
};

QWidget *createTasksPage(QWidget *parent) { return new TasksPage(parent); }

} // namespace sxcl::ui

#include "tasks_page.moc"
